src/ecalc/libraries/libecalc/common/tests/core/consumers/test_crossover_rates.py

```diff
@@ -0,0 +1,84 @@
+    expected_rates_within_capacity = [[2, 2], [2, 2]]
+
+    crossover_rate, rates_within_capacity = ConsumerSystem._get_crossover_rates(max_rate, rates)
+    assert np.array_equal(crossover_rate, expected_crossover_rate)
+    assert np.array_equal(rates_within_capacity, expected_rates_within_capacity)
+
+    # All rates within capacity
+    max_rate = [4, 4]
+    rates = [[3, 3], [1, 1]]
+    expected_crossover_rate = [0, 0]
+    expected_rates_within_capacity = [[3, 3], [1, 1]]
+
+    crossover_rate, rates_within_capacity = ConsumerSystem._get_crossover_rates(max_rate, rates)
+    assert np.array_equal(crossover_rate, expected_crossover_rate)
+    assert np.array_equal(rates_within_capacity, expected_rates_within_capacity)
+
+    # Exceeds capacity, cross over required
+    max_rate = [4, 4]  # 4 for both timesteps
+    rates = [[4, 4], [1, 1]]  # 4 in rate + 1 crossover, 1 exceeding
+    expected_crossover_rate = [1, 1]  # 1 exceeding
+    expected_rates_within_capacity = [[4, 4], [0, 0]]  # we have capacity for first stream, but not 2nd stream
+
+    crossover_rate, rates_within_capacity = ConsumerSystem._get_crossover_rates(max_rate, rates)
+    assert np.array_equal(crossover_rate, expected_crossover_rate)
+    assert np.array_equal(rates_within_capacity, expected_rates_within_capacity)
+
+    # Exceeds capacity, cross over required
+    max_rate = [4, 4]  # 4 for both timesteps
+    rates = [[5, 5]]  # 4 in rate
+    expected_crossover_rate = [1, 1]  # 1 exceeding
+    expected_rates_within_capacity = [[4, 4]]
+
+    crossover_rate, rates_within_capacity = ConsumerSystem._get_crossover_rates(max_rate, rates)
+    assert np.array_equal(crossover_rate, expected_crossover_rate)
+    assert np.array_equal(rates_within_capacity, expected_rates_within_capacity)
+
+    # Exceeds capacity, cross over required - fails as well..correct? leads to expected_rates_within_capacity = [[6, 6], [0, 0]]
+    # max_rate = [4, 4]  # 4 for both timesteps
+    # rates = [[5, 5], [2, 2]]  # 4 in rate
+    # expected_crossover_rate = [3, 3]  # 1 exceeding
+    # expected_rates_within_capacity = [[4, 4], [0, 0]]
+    #
+    # crossover_rate, rates_within_capacity = ConsumerSystem._get_crossover_rates(max_rate, rates)
+    # assert (np.array_equal(crossover_rate, expected_crossover_rate))
+    # assert (np.array_equal(rates_within_capacity, expected_rates_within_capacity))
+
+    # Exceeds capacity, cross over required - bug? leads to expected_rates_within_capacity = [[6, 6], [0, 0]]
+    # max_rate = [4, 4]  # 4 for both timesteps
+    # rates = [[5, 5], [0, 0]]  # 4 in rate
+    # expected_crossover_rate = [1, 1]  # 1 exceeding
+    # expected_rates_within_capacity = [[4, 4], [0, 0]]
+    #
+    # crossover_rate, rates_within_capacity = ConsumerSystem._get_crossover_rates(max_rate, rates)
+    # assert (np.array_equal(crossover_rate, expected_crossover_rate))
+    # assert (np.array_equal(rates_within_capacity, expected_rates_within_capacity))
```

Comment:
Seems like a bug to me

Reply:
Ok, so you agree that both my tests should succeeed?

Reply:
Yes